### ISSUE TITLE
takeHeapSnapshot timed out

### ISSUE THREAD

**Issue by Arooba-git** [facebook/memlab#107]
Hi :) Thanks for this handy tool!

This is more of a question than an issue: I have used memlab with over 40 projects to get the heap size and memory leaks without much problem 👍 . Recently I have come across cases where heap size exceeds 1 GB and I'm getting 'Page crashed!' along with the suggestion to ' Increase the 'protocolTimeout' setting in launch/connect calls for a higher timeout if needed.' I want to know where to set this timeout value..
Thanks!
<img width="1461" alt="Screen Shot 2024-01-12 at 9 49 28 PM" src="https://github.com/facebook/memlab/assets/56495631/f6e8f411-c3df-4f08-89ed-e8786d6f8fdd">

**Comment by JacksonGL** [comment]
Thanks for using memlab. I can increase the default protocol timeout (currently it's 180s based on puppeteer's default setting) and perhaps adding a new option to allow setting protocol timeout via CLI.

**Comment by Arooba-git** [comment]
Yes, that would be great, thanks. I would appreciate if you could notify me once it's done. :)

**Comment by Arooba-git** [comment]
@JacksonGL  Sorry to bother you again, but could I get an estimate of when the above issue will be resolved? Thanks.

**Comment by JacksonGL** [comment]
@Arooba-git I have a diff draft ready, but Monday is a holiday in the US (need my co-worker's review to ship it). So hopefully it will be landed on Tuesday. The code change will be linked to this GitHub issue, so GitHub should send a notification once the change is synced here.

In the meantime, if this is urgent for you, you can make this change [here](https://github.com/facebook/memlab/blob/ec905b1d50322ac70e6eb17bab4f679dd7d18a9a/packages/core/src/lib/Config.ts#L291) to your local git clone of memlab to unblock:

```typescript
this.puppeteerConfig = {
  protocolTimeout: 5 * 60 * 1000, // set timeout to be 5 minutes
  ...
}
```

Then add proper typing annotations and build the local copy

**Comment by JacksonGL** [comment]
@Arooba-git The default protocol timeout has been increased from 3 minutes to 5 minutes. You can also use the new CLI flag `--protocol-timeout <timeout-in-ms>` to further increase the timeout in memlab v1.1.43.

**Comment by Arooba-git** [comment]
@JacksonGL   Thanks for the prompt action. I'll try it out today 👍

**Comment by Arooba-git** [comment]
@JacksonGL Unfortunately I am getting the same error.. even after passing the --protocol-timeout with 300+ minutes (20000000 ms)  :(. If I provide you with the scenario files it is happening on, would you be able to perhaps assist in finding out the root cause since I don't have much experience with Puppeteer, and it seems like the library's internal issue?

<img width="478" alt="Screen Shot 2024-01-18 at 1 22 13 PM" src="https://github.com/facebook/memlab/assets/56495631/9b10d5c6-d75f-4b61-9467-4de04b51d036">

<img width="1241" alt="Screen Shot 2024-01-18 at 10 07 25 AM" src="https://github.com/facebook/memlab/assets/56495631/ab52ffea-7112-4995-bfe6-f7c291391e95">

<img width="1241" alt="Screen Shot 2024-01-18 at 10 08 24 AM" src="https://github.com/facebook/memlab/assets/56495631/d1cc23b5-0f05-42a4-8979-5961647727c0">

<img width="407" alt="Screen Shot 2024-01-18 at 10 13 26 AM" src="https://github.com/facebook/memlab/assets/56495631/f9a877c5-a085-4f0a-bcaa-b9a75f196a76">

**Comment by JacksonGL** [comment]
@Arooba-git I tested the new `--protocol-timeout` flag and it does timeout when I set the timeout to be very small, so the code change here indeed increased the protocol timeout.

The exception from your screenshot shows that it is thrown from Puppeteer's code; and, therefore it is outside of MemLab's scope. This is likely related to the connection between puppeteer and the headless browser. I would recommend filing a bug report on the Puppeteer's GitHub repo. 

My guess is that either 1) 1.4GB of memory is too much for Chrome to capture the heap snapshot (you could try taking the 1.4GB heap snapshot manually in Chrome DevTools to see if the Browser DevTools times out) or 2) the connection between Puppeteer and Chrome was not designed to handle such a large heap snapshot.

Moreover, if a web app uses more than 1GB JS heaps, it has some serious memory issues, I would recommend testing on a situation where the memory leak is less severe.

**Comment by Arooba-git** [comment]
Alright, I'll do that then. 👍  Thank you for the guidance! You may close this issue.

**Comment by JacksonGL** [comment]
MemLab disabled notification in browser by default:

To quickly see if may unblock, you can comment out this flag and build it locally: 
https://github.com/facebook/memlab/blob/73c9fa3addd87eb50bc13129224ee467ad895d4b/packages/core/src/lib/Config.ts#L304-L305

Feel free to open a separate issue about this.
